File: src/sprout/lexical.py
"""Pure-Python BM25 (Okapi) lexical index — the lexical half of hybrid retrieval.

Zero dependencies, fully deterministic. It tokenises with the *same* ``content_tokens``
as the dense embedder and the extractive generator, so a passage that ranks well
lexically is described by the same vocabulary the generator will quote and the judge
will check. BM25 catches exact-term matches (species names, "10 days") that a bag-of-
hashes dense vector can blur, so the two retrieval paths are genuinely complementary.

Internally this is a **term-major inverted index** (``term -> {doc_index: term_freq}``),
not a per-document scan table: scoring a query only ever touches documents that contain
at least one query term, via ``self._postings[term]``, rather than iterating every
document in the corpus. That is what makes it possible to persist the fitted index at
ingest time (``to_state``/``from_state``, FIX-07) and still have per-query cost track the
number of matching documents rather than the size of the corpus.
"""

from __future__ import annotations

import math
from collections import Counter
from collections.abc import Sequence
from typing import Any

from .text import content_tokens


class BM25Index:
    """Classic Okapi BM25 over a fixed set of documents (one per chunk)."""

    def __init__(self, documents: Sequence[str], *, k1: float = 1.5, b: float = 0.75) -> None:
        self.k1 = k1
        self.b = b
        tokenized = [content_tokens(d) for d in documents]
        self._lengths: list[int] = [len(toks) for toks in tokenized]
        n = len(tokenized)
        self._n = n
        self._avg_len = (sum(self._lengths) / n) if n else 0.0

        # Inverted index: term -> {doc_index: term_freq}. Built once per corpus; scoring
        # walks only the postings list of each query term, never every document.
        postings: dict[str, dict[int, int]] = {}
        df: Counter[str] = Counter()
        for i, toks in enumerate(tokenized):
            freq = Counter(toks)
            for term, tf in freq.items():
                postings.setdefault(term, {})[i] = tf
            df.update(freq.keys())
        self._postings = postings
        # Okapi idf with the +1 smoothing that keeps it non-negative.
        self._idf: dict[str, float] = {
            term: math.log(1 + (n - d + 0.5) / (d + 0.5)) for term, d in df.items()
        }

    def _sparse_scores(self, query: str) -> dict[int, float]:
        """BM25 score of ``query`` for every document it shares a term with (sparse).

        Iterates query terms outermost so that, for any document matched by more than
        one query term, contributions accumulate in the same order ``scores`` would
        produce them in — the two representations agree bit-for-bit, not just in rank.
        """
        q_terms = content_tokens(query)
        out: dict[int, float] = {}
        if not q_terms or self._avg_len == 0.0:
            return out
        for term in q_terms:
            docs = self._postings.get(term)
            if not docs:
                continue
            idf = self._idf.get(term, 0.0)
            for i, tf in docs.items():
                denom_norm = self.k1 * (1 - self.b + self.b * (self._lengths[i] / self._avg_len))
                out[i] = out.get(i, 0.0) + idf * (tf * (self.k1 + 1)) / (tf + denom_norm)
        return out

    def scores(self, query: str) -> list[float]:
        """BM25 score of ``query`` against every document, in document order."""
        out = [0.0] * self._n
        for i, score in self._sparse_scores(query).items():
            out[i] = score
        return out

    def ranking(self, query: str) -> list[int]:
        """Document indices ordered best-first, dropping zero-score documents.

        Cost is proportional to the number of documents sharing a query term (the union
        of their postings lists), not the corpus size — see the module docstring.
        """
        scored = [(i, s) for i, s in self._sparse_scores(query).items() if s > 0.0]
        scored.sort(key=lambda pair: (-pair[1], pair[0]))
        return [i for i, _ in scored]

    # --- persistence -------------------------------------------------------------
    # The postings (inverted term -> {doc: tf}, idf, lengths, avg length) are exactly
    # what ``scores``/``ranking`` need; persisting them lets ``from_state`` reconstruct
    # an index without re-tokenising every document, which is the per-query cost this
    # class exists to avoid once the corpus is indexed at ingest time (FIX-07).
    def to_state(self) -> dict[str, object]:
        """Serialise the fitted postings — everything needed to score without retokenising."""
        return {
            "k1": self.k1,
            "b": self.b,
            "n": self._n,
            "avg_len": self._avg_len,
            "idf": dict(self._idf),
            "lengths": list(self._lengths),
            "postings": {
                term: {str(i): tf for i, tf in docs.items()}
                for term, docs in self._postings.items()
            },
        }

    @classmethod
    def from_state(cls, state: dict[str, Any]) -> BM25Index:
        """Reconstruct an index from ``to_state`` output — no document retokenisation."""
        self = cls.__new__(cls)
        self.k1 = float(state["k1"])
        self.b = float(state["b"])
        self._n = int(state["n"])
        self._avg_len = float(state["avg_len"])
        self._idf = {str(term): float(v) for term, v in state["idf"].items()}
        self._lengths = [int(x) for x in state["lengths"]]
        self._postings = {
            str(term): {int(i): int(tf) for i, tf in docs.items()}
            for term, docs in state["postings"].items()
        }
        return self

File: web-static/src/lexical.ts
/**
 * Pure-TypeScript BM25 (Okapi) lexical index — a mirror of `src/sprout/lexical.py`.
 *
 * Zero dependencies, fully deterministic, tokenises with the same `contentTokens` as the
 * dense embedder and the extractive generator (see `text.ts`).
 *
 * Mirrors FIX-07's inverted-postings layout (`term -> {doc_index: term_freq}`): the index
 * is built (or loaded from the postings `sprout ingest` persisted in `index.json`) once
 * per corpus, and scoring walks only the postings list of each query term. Contribution
 * accumulation is query-term-outermost, exactly like Python's `_sparse_scores`, so the
 * two implementations agree bit-for-bit, not just in rank.
 */

import { contentTokens } from "./text.js";

/** Shape of the postings `store.py`'s `BM25Index.to_state()` persists in `index.json`. */
export interface BM25State {
  k1: number;
  b: number;
  n: number;
  avg_len: number;
  idf: Record<string, number>;
  lengths: number[];
  postings: Record<string, Record<string, number>>;
}

export class BM25Index {
  readonly k1: number;
  readonly b: number;

  private readonly lengths: number[];
  private readonly n: number;
  private readonly avgLen: number;
  private readonly idf: Map<string, number>;
  // term -> (doc index -> term frequency), insertion-ordered by ascending doc index.
  private readonly postings: Map<string, Map<number, number>>;

  private constructor(
    k1: number,
    b: number,
    n: number,
    avgLen: number,
    lengths: number[],
    idf: Map<string, number>,
    postings: Map<string, Map<number, number>>,
  ) {
    this.k1 = k1;
    this.b = b;
    this.n = n;
    this.avgLen = avgLen;
    this.lengths = lengths;
    this.idf = idf;
    this.postings = postings;
  }

  static build(documents: readonly string[], k1 = 1.5, b = 0.75): BM25Index {
    const tokenized = documents.map((d) => contentTokens(d));
    const lengths = tokenized.map((toks) => toks.length);
    const n = tokenized.length;
    const totalLen = lengths.reduce((a, b2) => a + b2, 0);
    const avgLen = n ? totalLen / n : 0.0;

    const postings = new Map<string, Map<number, number>>();
    const df = new Map<string, number>();
    for (let i = 0; i < tokenized.length; i++) {
      const freq = new Map<string, number>();
      for (const tok of tokenized[i] as string[]) {
        freq.set(tok, (freq.get(tok) ?? 0) + 1);
      }
      for (const [term, tf] of freq) {
        let docs = postings.get(term);
        if (docs === undefined) {
          docs = new Map<number, number>();
          postings.set(term, docs);
        }
        docs.set(i, tf);
        df.set(term, (df.get(term) ?? 0) + 1);
      }
    }
    const idf = new Map<string, number>();
    for (const [term, d] of df) {
      idf.set(term, Math.log(1 + (n - d + 0.5) / (d + 0.5)));
    }
    return new BM25Index(k1, b, n, avgLen, lengths, idf, postings);
  }

  /** Reconstruct an index from the postings `sprout ingest` persisted (`to_state` mirror). */
  static fromState(state: BM25State): BM25Index {
    const idf = new Map<string, number>();
    for (const [term, v] of Object.entries(state.idf)) {
      idf.set(term, v);
    }
    const postings = new Map<string, Map<number, number>>();
    for (const [term, docs] of Object.entries(state.postings)) {
      const parsed = new Map<number, number>();
      for (const [i, tf] of Object.entries(docs)) {
        parsed.set(Number(i), tf);
      }
      postings.set(term, parsed);
    }
    return new BM25Index(
      state.k1,
      state.b,
      state.n,
      state.avg_len,
      [...state.lengths],
      idf,
      postings,
    );
  }

  /**
   * BM25 score of `query` for every document it shares a term with (sparse). Query terms
   * iterate outermost so multi-term contributions accumulate in the same order as the
   * Python `_sparse_scores` — bit-for-bit agreement, not just rank agreement.
   */
  private sparseScores(query: string): Map<number, number> {
    const qTerms = contentTokens(query);
    const out = new Map<number, number>();
    if (qTerms.length === 0 || this.avgLen === 0.0) {
      return out;
    }
    for (const term of qTerms) {
      const docs = this.postings.get(term);
      if (docs === undefined || docs.size === 0) {
        continue;
      }
      const idf = this.idf.get(term) ?? 0.0;
      for (const [i, tf] of docs) {
        const length = this.lengths[i] as number;
        const denomNorm = this.k1 * (1 - this.b + this.b * (length / this.avgLen));
        out.set(i, (out.get(i) ?? 0.0) + (idf * (tf * (this.k1 + 1))) / (tf + denomNorm));
      }
    }
    return out;
  }

  /** BM25 score of `query` against every document, in document order. */
  scores(query: string): number[] {
    const out = new Array<number>(this.n).fill(0.0);
    for (const [i, score] of this.sparseScores(query)) {
      out[i] = score;
    }
    return out;
  }

  /** Document indices ordered best-first, dropping zero-score documents. */
  ranking(query: string): number[] {
    const scored: [number, number][] = [];
    for (const [i, s] of this.sparseScores(query)) {
      if (s > 0.0) {
        scored.push([i, s]);
      }
    }
    scored.sort((a, b2) => (b2[1] - a[1] !== 0 ? b2[1] - a[1] : a[0] - b2[0]));
    return scored.map(([i]) => i);
  }
}
